File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    n = len(maps)
    m = len(maps[0])
        
    dx = [0,0,1,-1]
    dy = [1,-1,0,0]
    
    q = deque([(0,0)])
    while q:
        item = q.popleft()
        x, y = item[0], item[1]
        
        for i in range(4):
            nx, ny = x + dx[i], y + dy[i]
            
            if nx < 0 or nx > n-1 or ny < 0 or ny > m-1:
                continue
                
            if maps[nx][ny] != 0 and maps[nx][ny] == 1:
                maps[nx][ny] = maps[x][y] + 1
                q.append((nx, ny))
                
                if maps[-1][-1] != 1:
                    return maps[-1][-1]
                
    return -1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    public int solution(int[][] maps) {
        int n = maps.length;
        int m = maps[0].length;
        int answer = 0;
        
        int[] dx = new int[]{1, -1, 0, 0};
        int[] dy = new int[]{0, 0, 1, -1};
        
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{0,0});
        
        while (!q.isEmpty()) {
            int[] item = q.poll();
            int x = item[0];
            int y = item[1];
            
            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];
                
                if (nx < 0 || nx >= n || ny < 0 || ny >= m) {
                    continue;    
                }
                
                if (maps[nx][ny] != 0 && maps[nx][ny] == 1) {
                    maps[nx][ny] = maps[x][y] + 1;
                    q.offer(new int[]{nx, ny});
                    
                    if (maps[n-1][m-1] != 1) {
                        return maps[n-1][m-1];
                    }
                }
            }
        }
        return -1;
    }
}